docs/modules/SDDs/pages/0021-cluster-catalog-compilation.adoc

```diff
@@ -0,0 +1,145 @@
+* Solution to automate the component updates
+
+
+== Design Proposal
+
+The design splits the problem into two parts:
+
+. Figuring out which cluster catalogs need to be compiled
+. Running the actual compilation job
+
+
+=== Quartermaster
+
+A new tool called https://en.wikipedia.org/wiki/Quartermaster#Royal_Navy[Quartermaster] is implemented.
+This tool should solve the problem of figuring out which cluster catalogs need to be recompiled.
+The tool gets input in the form of webhooks which are triggered by Git repositories or other tools.
+By using the Kapitan inventory the tool needs to figure out for which clusters the config changed and therefore needs a recompilation.
+In a first step this could be relatively rudimentary and only use the input information from a webhook (cluster/tenant ID) to decide which clusters to compile.
+
+The information if a cluster needs to be compiled is then stored in the cluster object of the Lieutenant API.
+In a first PoC phase this can be an annotation like `cluster.syn.tools/needs-compilation=true`.
+In a later stage this information should be part of the typed struct and might include further information like the exact version of the various input repos.
+
+
+==== Webhooks
+
+Triggering a webhook can be implemented in either a Git repo natively (for example https://developer.github.com/webhooks/[GitHub Webhooks]) or via a CI/CD system (for example a https://docs.gitlab.com/ee/ci/[GitLab CI] job which runs `curl`).
+A webhook can also be called by any other tool or even manually with a `curl` command.
+
+A webhook can include the following optional information:
+
+* Cluster ID for which to compile the catalog.
+This can be used in case where a specific cluster should be compiled
+
+* Tenant ID for which to figure out affected clusters.
+This can be used to limit the set of clusters to be considered
+
+* Set of classes that changed
+
+
+==== Authentication & Authorization
+
+To prevent exsessive resource usage or even DoS attacks, the webhooks need to be authenticated.
+A permission system needs to be in place to decide which entities are allowed to trigger which actions.
+It shouldn't be possible for example for a tenant to trigger compilation for clusters of other tenants.
+
+
+=== Compilation Jobs
+
+To compile the catalog for a cluster, a compilation job is created.
+The compilation job is a https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion[Kubernetes job] resource which executes the xref:0006-commodore-configuration-catalog-compilation.adoc[catalog compilation engine].
+
+To distribute the load this compilation generates and to follow a decentralized approach, these jobs are run on the same cluster as they compile the catalog for.
+The Steward component on a cluster starts a compilation job if the `cluster.syn.tools/needs-compilation` annotation is set to `true` on its own cluster object.
+It will then launch the respective job to compile the cluster catalog.
```

Comment:
I would like to see at least a sketch of how Steward will launch the job to compile the catalog in this SDD.

Right now we're using argo-events to create plain Kubernetes jobs to compile catalogs. I don't think it's necessary to have argo-events on each cluster, but we probably want to have some compile job template that can be instantiated by Steward when a catalog needs to be compiled.

Reply:
I was thinking that Steward would just create the Job via the K8s API (so the "template" would be in go code)

Reply:
How do we find out when the compile job fails? Maybe we need some Prometheus rules explicitly for that.?

Reply:
Good point.
We could push metrics to the Prom pushgateway from the compile job itself.

Reply:
I added a paragraph for each inventory parsing and metrics respectively.